### PR TITLE
Line sensor/follower

### LINKED ISSUE #18: [ROBOT] Line followers
We want to add sensor attachments for the robot, to encourage students to use these in the real competition. The first two sensors we should implement are:
- Line followers: can detect transitions from light to dark (lines)
- Ultrasonic distance sensors: tells the distance away from objects

### PR DESCRIPTION
**Pull request by kevinzhu00**

### CODE REVIEW

**Comment by charleshong3** [comment]
Looks like you need to `git pull origin master` and fix merge conflicts

File: static/js/robot.js [hunk 1]
```diff
@@ -92,7 +103,7 @@ class RobotClass {
             Y: this.Y,
             dir: this.dir
         };
-
+        this.sensor.get_val()
```
Comment (charleshong3) [inline static/js/robot.js:107]: Do we need to get the sensor values every time we update position, or can we limit it to when the student code requests it?

File: static/js/robot.js [hunk 2]
```diff
@@ -40,21 +40,32 @@ class RobotClass {
...
       this.runningCoroutines = new Set();
+
+      // Ensure we don't hit sync errors when updating our values
+      this.leftSensor = 0;
+      this.centerSensor = 0;
+      this.rightSensor = 0;
+      this.autonomous_main = autonomous_main
+      this.sensor = new Sensor(this);
+      this.tapeLines = [];
+      this.tapeLines.push(new TapeLine(27, 27, 115, 115));
```
Comment (charleshong3) [inline static/js/robot.js]: We should probably read this in from a file, and align it with how the walls/collidable objects are read in (collaborate with Justin/Paul on this). However, I would probably be ok with doing this in a future PR.

File: static/js/robot.js [hunk 3]
```diff
@@ -40,21 +40,32 @@ class RobotClass {
...
-      this.ltheta = 0.0;        // angular position of l wheel, degrees
-      this.rtheta = 0.0;        // angular position of r wheel, degrees
-      this.dir = 0.0;           // Direction of the robot facing, degrees
+    neg = -1;                    // negate left motor calculation
+    startX = 70.0
+    startY = 70.0
+
+    constructor(autonomous_main=null) {
+      this.X = this.startX;           // X position of the robot
+      this.Y = this.startY;           // Y position of the robot
```
Comment (charleshong3) [inline static/js/robot.js]: I think these are the old comments, let's add the new comments back (see diff)

File: static/js/base.js [hunk 4]
```diff
@@ -64,14 +64,23 @@ function update(state) {
...
     const scaleFactor = 3;
-    const scaledX = state.X * scaleFactor - 30;
-    const scaledY = state.Y * scaleFactor - 40;
+    const centerX = state.X * scaleFactor;
+    const centerY = state.Y * scaleFactor;
     const dir = state.dir;
     document.getElementById("demo").innerHTML = state.X.toFixed(2) + ", " + state.Y.toFixed(2)
+    const sensorPoints = document.querySelectorAll("circle")
+    const scaledX = centerX - 30
+    const scaledY = centerY - 40
```
Comment (charleshong3) [inline static/js/base.js:73]: Changing scaledX and scaledY then reusing the variable names for something different probably isn't best practice. I would recommend changed scaledX and scaledY to something else, ideally something that references the fact that they refer to the top left corner of the robot.

**Comment by charleshong3** [comment]
We also need to add the line following Python code somewhere!

File: static/js/robot.js [hunk 5]
```diff
@@ -479,7 +503,7 @@ class Simulator{
         this.autonomous_main = env['autonomous_main']
         this.teleop_setup = env['teleop_setup']
         this.teleop_main = env['teleop_main']
-
+        console.log(env['autonomous_main'])
```
Comment (charleshong3) [inline static/js/robot.js]: Delete this line

File: static/js/robot.js [hunk 6]
```diff
@@ -445,6 +468,7 @@ class Simulator{
         this.mode = "idle";
         this.initGamepad();
         this.current = [];
+        this.tapelines = [];
```
Comment (charleshong3) [inline static/js/robot.js]: Do we use this? tapeLines is a RobotClass attribute, right?

File: static/js/robot.js [hunk 7]
```diff
@@ -594,3 +618,90 @@ this.onmessage = function(e) {
...
+             let distX = Math.abs(sensor_x-x)
+             let distY = Math.abs(sensor_y-y)
+             let distSquared = (distX*distX)+(distY*distY)
+             totalLine += Math.min(1,3/distSquared)
+           }
+         }
+       }
+       total.push(Math.min(totalLine, 1))
+     }
+     console.log(total)
```
Comment (charleshong3) [inline static/js/robot.js]: We should delete this line. We can maybe add a Sensors tab (next to Code and Gamepad) that lists all current sensor readings. If we do this, we will actually have to read all the sensors every tick, nullifying my comment above. If this affects the performance of the simulator, we could look into doing something like updating the sensor readings every other tick, or every 5 ticks.

File: static/js/robot.js (hunk 1 above)
Comment (charleshong3) [reply static/js/robot.js:107]: Maybe we should get sensor readings every tick. See https://github.com/pioneers/pimulator-js/pull/21#discussion_r536916145

File: static/js/robot.js [hunk 8]
```diff
@@ -40,21 +40,32 @@ class RobotClass {
...
-      this.Wl = 0.0;            // angular velocity of l wheel, radians/s
-      this.Wr = 0.0;            // angular velocity of r wheel, radians/s
-      this.ltheta = 0.0;        // angular position of l wheel, degrees
-      this.rtheta = 0.0;        // angular position of r wheel, degrees
-      this.dir = 0.0;           // Direction of the robot facing, degrees
+    neg = -1;                    // negate left motor calculation
+    startX = 70.0
+    startY = 70.0
+
+    constructor(autonomous_main=null) {
```
Comment (charleshong3) [inline static/js/robot.js]: I don't think we're actually using this argument anymore, can this be confirmed?

File: static/js/base.js (hunk 4 above)
Comment (charleshong3) [inline static/js/base.js]: I know scaledX and scaledY are the names we were using before, but something like topLeftX and topLeftY may make more sense.

File: static/js/robot.js (hunk 2 above)
Comment (kevinzhu00) [reply static/js/robot.js]: For the future :)

**Review by charleshong3 — APPROVED** [review]
lgtm

**Pull request merged**